File: sdks/go/autonoma/token_resolution_test.go
package autonoma

import (
	"errors"
	"reflect"
	"strings"
	"testing"
)

func TestResolveTokensTestRunID(t *testing.T) {
	out, err := ResolveTokens(map[string]any{"email": "alice-{{testRunId}}@test.local"}, "run-123", 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	expected := map[string]any{"email": "alice-run-123@test.local"}
	if !reflect.DeepEqual(out, expected) {
		t.Fatalf("got %v, want %v", out, expected)
	}
}

func TestResolveTokensIndex(t *testing.T) {
	out, err := ResolveTokens(map[string]any{"slot": "pos-{{index}}"}, "r", 4)
	if err != nil {
		t.Fatal(err)
	}
	expected := map[string]any{"slot": "pos-4"}
	if !reflect.DeepEqual(out, expected) {
		t.Fatalf("got %v, want %v", out, expected)
	}
}

func TestResolveTokensCycleWraps(t *testing.T) {
	cases := []struct {
		index int
		want  string
	}{{0, "a"}, {1, "b"}, {2, "a"}}
	for _, c := range cases {
		out, err := ResolveTokens("{{cycle(a,b)}}", "r", c.index)
		if err != nil {
			t.Fatal(err)
		}
		if out != c.want {
			t.Fatalf("index=%d: got %v, want %v", c.index, out, c.want)
		}
	}
}

func TestResolveTokensCycleQuoted(t *testing.T) {
	out, err := ResolveTokens("{{cycle('WEB','IOS','ANDROID')}}", "r", 1)
	if err != nil {
		t.Fatal(err)
	}
	if out != "IOS" {
		t.Fatalf("got %v, want IOS", out)
	}
}

func TestResolveTokensNested(t *testing.T) {
	input := map[string]any{
		"users": []any{
			map[string]any{"email": "u-{{testRunId}}@t.local"},
			map[string]any{"email": "v-{{testRunId}}@t.local"},
		},
		"tags": []any{"{{testRunId}}-a", "{{testRunId}}-b"},
	}
	expected := map[string]any{
		"users": []any{
			map[string]any{"email": "u-xyz@t.local"},
			map[string]any{"email": "v-xyz@t.local"},
		},
		"tags": []any{"xyz-a", "xyz-b"},
	}
	out, err := ResolveTokens(input, "xyz", 0)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(out, expected) {
		t.Fatalf("got %v, want %v", out, expected)
	}
}

func TestResolveTokensMultipleInOneString(t *testing.T) {
	out, err := ResolveTokens("{{testRunId}}-{{index}}", "run", 7)
	if err != nil {
		t.Fatal(err)
	}
	if out != "run-7" {
		t.Fatalf("got %v, want run-7", out)
	}
}

func TestResolveTokensUnknownRaises(t *testing.T) {
	_, err := ResolveTokens(map[string]any{"x": "hello-{{mystery}}"}, "r", 0)
	if err == nil {
		t.Fatal("expected error")
	}
	var ae *AutonomaError
	if !errors.As(err, &ae) {
		t.Fatalf("expected AutonomaError, got %T: %v", err, err)
	}
	if ae.Code != "UNRESOLVED_TOKEN" {
		t.Fatalf("expected UNRESOLVED_TOKEN, got %s", ae.Code)
	}
	if !strings.Contains(ae.Message, "mystery") {
		t.Fatalf("expected message to contain mystery, got %q", ae.Message)
	}
}

func TestResolveTokensPrimitivesPassThrough(t *testing.T) {
	for _, v := range []any{42, true, nil} {
		out, err := ResolveTokens(v, "r", 0)
		if err != nil {
			t.Fatal(err)
		}
		if !reflect.DeepEqual(out, v) {
			t.Fatalf("got %v, want %v", out, v)
		}
	}
}

func TestResolveTokensStringWithoutTokens(t *testing.T) {
	out, err := ResolveTokens("plain string", "r", 0)
	if err != nil {
		t.Fatal(err)
	}
	if out != "plain string" {
		t.Fatalf("got %v, want plain string", out)
	}
}

File: sdks/java/autonoma-sdk/src/test/java/ai/autonoma/sdk/TokenResolutionTest.java
package ai.autonoma.sdk;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class TokenResolutionTest {

    @Test
    @SuppressWarnings("unchecked")
    void testRunIdSubstituted() {
        Map<String, Object> input = new LinkedHashMap<>();
        input.put("email", "alice-{{testRunId}}@test.local");
        Map<String, Object> out = (Map<String, Object>) AutonomaHandler.resolveTokens(input, "run-123", 0);
        assertEquals("alice-run-123@test.local", out.get("email"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void indexSubstituted() {
        Map<String, Object> input = new LinkedHashMap<>();
        input.put("slot", "pos-{{index}}");
        Map<String, Object> out = (Map<String, Object>) AutonomaHandler.resolveTokens(input, "r", 4);
        assertEquals("pos-4", out.get("slot"));
    }

    @Test
    void cycleSubstitutedAndWraps() {
        assertEquals("a", AutonomaHandler.resolveTokens("{{cycle(a,b)}}", "r", 0));
        assertEquals("b", AutonomaHandler.resolveTokens("{{cycle(a,b)}}", "r", 1));
        assertEquals("a", AutonomaHandler.resolveTokens("{{cycle(a,b)}}", "r", 2));
    }

    @Test
    void cycleQuotedValuesStripped() {
        assertEquals("IOS", AutonomaHandler.resolveTokens("{{cycle('WEB','IOS','ANDROID')}}", "r", 1));
    }

    @Test
    @SuppressWarnings("unchecked")
    void nestedStructuresWalked() {
        Map<String, Object> user1 = new LinkedHashMap<>();
        user1.put("email", "u-{{testRunId}}@t.local");
        Map<String, Object> user2 = new LinkedHashMap<>();
        user2.put("email", "v-{{testRunId}}@t.local");
        Map<String, Object> input = new LinkedHashMap<>();
        input.put("users", List.of(user1, user2));
        input.put("tags", List.of("{{testRunId}}-a", "{{testRunId}}-b"));

        Map<String, Object> out = (Map<String, Object>) AutonomaHandler.resolveTokens(input, "xyz", 0);
        List<Map<String, Object>> users = (List<Map<String, Object>>) out.get("users");
        assertEquals("u-xyz@t.local", users.get(0).get("email"));
        assertEquals("v-xyz@t.local", users.get(1).get("email"));
        List<String> tags = (List<String>) out.get("tags");
        assertEquals(List.of("xyz-a", "xyz-b"), tags);
    }

    @Test
    void multipleTokensInOneString() {
        assertEquals("run-7", AutonomaHandler.resolveTokens("{{testRunId}}-{{index}}", "run", 7));
    }

    @Test
    void unknownTokenRaises() {
        Map<String, Object> input = new LinkedHashMap<>();
        input.put("x", "hello-{{mystery}}");
        AutonomaError err = assertThrows(AutonomaError.class, () ->
            AutonomaHandler.resolveTokens(input, "r", 0));
        assertEquals("UNRESOLVED_TOKEN", err.getCode());
        assertTrue(err.getMessage().contains("mystery"));
    }

    @Test
    void nonStringPrimitivesPassThrough() {
        assertEquals(42, AutonomaHandler.resolveTokens(42, "r", 0));
        assertEquals(Boolean.TRUE, AutonomaHandler.resolveTokens(Boolean.TRUE, "r", 0));
        assertNull(AutonomaHandler.resolveTokens(null, "r", 0));
    }

    @Test
    void stringWithoutTokensUnchanged() {
        assertEquals("plain string", AutonomaHandler.resolveTokens("plain string", "r", 0));
    }
}
